### ISSUE TITLE
offline archive

### ISSUE THREAD

**Issue by jwinarske** [dart-lang/pub#4353]
I have a need to support Long Term Support (LTS); automotive and IOT.

I'm proposing a new sub command called archive. 

This would cache source types hosted and git into an archive folder. 

Then without Internet access you can restore said archive, and build.

I have an implementation in Python that does this today.  I am using this with the Yocto build system:
https://github.com/meta-flutter/meta-flutter/blob/scarthgap/tools/pubspec.py

**Comment by sigurdm** [comment]
We don't have a separate command for this today, but you can already do:

```
PUB_CACHE=.pub_cache dart pub get
```

Which would put the cache inside the current directory.
This way you can zip everything up.

Then when extracting somewhere run
```
PUB_CACHE=.pub_cache dart pub get --offline
```

To fix the absolute references in the .dart_tool/package_config.json

**Comment by jwinarske** [comment]
The use case is building 100+ apps simultaneously with LTS support. The proposed archive method dramatically decreases build storage.  As there is only one archive of a package, which may show up 100+ times.  The compress pub cache approach use much more disk space at scale.  The pub cache restore step takes ~1 second; multi-threaded.

**Comment by jwinarske** [comment]
@sigurdm
Archive would not include any redundant packages; only single references for git and hosted artifacts.

1. pub archive <path>
Referencing the pubspec.lock file fetch hosted tar.gz, create tar.gz.256, and create .cache folder with <package>-advisories.json, and <package>-version.json files.  This represents the archive; all the items fetched from the pub server.

2. pub archive <path> restore
Referencing the pubspec.lock creates a pub cache using archive path.  Restores pub cache to path specified by PUB_CACHE.

**Comment by sigurdm** [comment]
> Archive would not include any redundant packages;

Not sure what you mean by this. The procedure I outlined above would only fetch the packages needed.

>  only single references for git and hosted artifacts.

What are "references" in this context?

**Comment by jwinarske** [comment]
> Not sure what you mean by this. The procedure I outlined above would only fetch the packages needed.

Yes this is true for a single application.  If I look at all the artifacts fetched for 100 apps, there are many common packages shared between them.  This means the archive folder would only contain one copy of each package version.  Using a compressed pub archive duplicates the same package version many times; thus increasing disk space.

> What are "references" in this context?

These would be considered a reference:
* hosted/pub.dev/package-x.x.x.tar.gz
  hosted/pub.dev/.cache/package-versions.json
  hosted/pub.dev/.cache/package-advisories.json
* git mirror
  git/cache/repo-sha1

**Comment by jwinarske** [comment]
Archive listing for [flutter-gallery](https://github.com/flutter/gallery)
[archive_example.txt](https://github.com/user-attachments/files/16694925/archive_example.txt)
